### ISSUE TITLE
Improved Application and Data Logging 

### ISSUE THREAD

**Issue by mstruve** [forem/forem#4926]
This is tied pretty closely to [Consolidating and Improving Application Monitoring](https://github.com/thepracticaldev/dev.to/issues/4925) but I believe can stand on its own as an issue.

In an effort to improve reliability we want to improve our application monitoring. One of the easiest ways to improve application monitoring is through logging. Currently, we are making use of the Rails default logs and loggers that come packaged with our gems. These are great, but not quite tailored for our needs. Ideally, it would be nice to have a logging system/template within dev that we can use to log events that we care about. Things like job runtimes, how long a job sits in the queue before it's worked on. How long it takes from a request being made to us updating the data. The list goes on. 

By tracking things like this we can monitor them and report on any irregularities we see. Reporting on stats like these can help us spot problems faster. It can also help us track requests and in the event of errors backtrace through events to better understand what caused the problem. Logging is also a great way to track data trends over time and the growth of data sets such as users, articles, comments, etc. Did we see a giant spike in User count overnight? That might indicate either crazy growth or a possible problem. 

Couple things to keep in mind when setting up logs, we want to have a set of keys that is consistent across all logs. This allows us to filter and sort logs more easily. If you have a `request_id` in one log but in another its called `req_id` you will have a harder time filtering those.

**Comment by delbetu** [comment]
Having more clarification on this would be great.
I think there are two different units of work here.
1. All the events that need to be logged.
2. The service that receives the logs and saves them somewhere.

For the first point, each event needs to be specified
For the second I would suggest using an external service such as sumologic.
The main feature of sumologic is its query language over the generated logs, it is flexible

In the other hand information about jobs execution stats, it seems that delayed_job already logs some measurements see https://github.com/tobi/delayed_job/blob/719b628bdd54566f80ae3a99c4a02dd39d386c07/lib/delayed/job.rb#L96

If you can identify which of the measurements are useful we could check if they already exist or figure out where to modify.

**Comment by mstruve** [comment]
Hey @delbetu!

You are definitely right this needs a lot more clarification before it is worked on. However, after our recent tech meeting, we are putting this in the "Future" work category as there are some things that still need to be sorted out beforehand such as what needs to be logged and where it is all going. Even though we are not currently working on it, I still thought it was important to create the issue so that the community can see what we are planning and there is better transparency into our process. Plus, it's nice to have it somewhere besides Notes on my computer 😉 

I am currently in the process of getting our [monitoring infrastructure all set up](https://github.com/thepracticaldev/dev.to/issues/4925) and once that is done then I see this as the next step. I fully plan to break this down into a lot of smaller issues once I have the logging framework in place so that others can contribute to the effort.  Definitely follow this issue if you want to stay updated on it as I move it forward.

Let me know if you have any other questions!

**Comment by ssimontis** [comment]
This sounds like a task for a good APM utility. I have played around with DataDog a little bit and it was pretty impressive, but it costs money. Any of the APM systems will be an investment, but they are a lot more effective than trying to guess which events you will need and implement everything yourself.

**Comment by mstruve** [comment]
Given our recent move to Datadog logging I think we can close this issue for now until we decide to upgrade our logging solution to something different such as Fluentd.
